File: answers/2/main.py
import numpy as np
import cv2

def bgr2gray(img):
    _img = img.copy().astype(np.float32)
    gray = _img[..., 0] * 0.2126 + _img[..., 1] * 0.7152 + _img[..., 2] * 0.0722
    gray = np.clip(gray, 0, 255) # ndarrayを[0,255]に収める
    return gray.astype(np.uint8)

if __name__ == '__main__':
    img_orig = cv2.imread('questions/dataset/images/imori_256x256.png')
    img_result = bgr2gray(img_orig)
    cv2.imshow('python_origin', img_orig)
    cv2.imshow('python_result', img_result)
    cv2.moveWindow('python_result', img_orig.shape[1], 0)
    cv2.waitKey(3000)

File: answers/2/main.cpp
#include <vector>
#include <chrono>
#include <iostream>
#include <opencv2/opencv.hpp>

void bgr2gray(const cv::Mat &src, cv::Mat &dst) {
    std::vector<cv::Mat> planes;
    cv::split(src, planes);
    dst = planes[0] * 0.0722 + planes[1] * 0.7152 + planes[2] * 0.2126;
    cv::threshold(dst, dst, 0, 255, cv::THRESH_TOZERO);
    cv::threshold(dst, dst, 255, 255, cv::THRESH_TRUNC);
    dst.convertTo(dst, CV_8UC1);
}

int main() {
    cv::Mat img_orig = cv::imread("questions/dataset/images/imori_256x256.png");
    cv::Mat img_result;
    bgr2gray(img_orig, img_result);

    cv::imshow("cpp_origin", img_orig);
    cv::imshow("cpp_result", img_result);
    cv::moveWindow("cpp_result", img_orig.size().width, 0);
    cv::waitKey(3000);
    return 0;
}
